content/en/blog/_posts/2023-05-05-memory-qos-cgroups-v2/index.md

```diff
@@ -0,0 +1,286 @@
+      Please refer to the <a href="https://github.com/kubernetes/enhancements/tree/master/keps/sig-node/2570-memory-qos">KEP</a>
+      for more details on the formula.</i>
+       </td>
+   </tr>
+</table>
+
+**Note** `memory.high` is set only on container level cgroups while `memory.min` is set on
+container, pod and node level cgroups.
+
+### `memory.min` calculations for cgroups heirarchy
+
+When container memory requests are made, kubelet passes `memory.min` to the back-end 
+CRI runtime (possibly containerd, cri-o) via the `Unified` field in CRI during 
+container creation. The `memory.min` in container level cgroups will be set to:  
+
+$memory.min =  pod.spec.containers[i].resources.requests[memory]$  
+<sub>for every i<sup>th</sup> container in a pod</sub>
+
+Since the `memory.min` interface requires that the ancestor cgroups directories are all 
+set, the pod and node cgroups directories need to be set correctly. 
+
+`memory.min` in pod level cgroup:  
+$memory.min = \sum_{i=0}^{no. of pods}pod.spec.containers[i].resources.requests[memory]$  
+
+`memory.min` in node level cgroup:  
+$memory.min = \sum_{i}^{no. of nodes}\sum_{j}^{no. of pods}pod[i].spec.containers[j].resources.requests[memory]$  
+
+Kubelet will manage the cgroups hierarchy of the pod level and node level cgroups 
+directly using the libcontainer library (from the runc project), while container
+cgroups limits are managed by the container runtime.
+
+### Support for Pod QoS classes
+
+Based on user feedback for the Alpha feature in Kubernetes v1.22, some users would like 
+to opt out of MemoryQoS on a per-pod basis to ensure there is no early memory throttling.
+Therefore, in Kubernetes v1.27 Memory QOS also supports memory.high to be set as per 
+Quality of Service(QoS) for Pod classes. Following are the different cases for memory.high
+as per QOS classes:
+
+1. **Guaranteed pods** by their QoS definition require memory requests=memory limits and are 
+not overcommitted. Hence MemoryQoS feature is disabled on those pods by not setting 
+memory.high. This ensures that Guaranteed pods can fully use their memory requests up
+to their set limit, and not hit any throttling.
+
+2. **Burstable pods** by their QoS definition require at least one container in the Pod with
+CPU or memory request or limit set.
+    
+    * When requests.memory and limits.memory are set, the formula is used as-is:
+
+    {{< figure src="/blog/2023/05/05/qos-memory-resources/container-memory-high-limit.svg" title="`memory.high` when requests and limits are set" alt="`memory.high` when requests and limits are set" >}}
+
+    * When requests.memory is set, limits.memory is not set, we substitute limits.memory
+    for node allocatable memory in the formula:
+
+    {{< figure src="/blog/2023/05/05/qos-memory-resources/container-memory-high-no-limits.svg" title="`memory.high` when requests and limits are not set" alt="`memory.high` when requests and limits are not set" >}}
```

Comment:
```suggestion
   * When requests.memory and limits.memory are set, the formula is used as-is:

     {{< figure src="/blog/2023/05/05/qos-memory-resources/container-memory-high-limit.svg" title="`memory.high` when requests and limits are set" alt="`memory.high` when requests and limits are set" >}}

   * When requests.memory is set, limits.memory is not set, we substitute limits.memory
     for node allocatable memory in the formula:

     {{< figure src="/blog/2023/05/05/qos-memory-resources/container-memory-high-no-limits.svg" title="`memory.high` when requests and limits are not set" alt="`memory.high` when requests and limits are not set" >}}
```

Reply:
be careful about the indentation. 4-spaces indentation sometimes confuses the parser. The parser may treat it as code snippets. For numbered lists, please always indent by 3-spaces.